### ISSUE TITLE
Late-injected options of constructors are dropped in resolveComponentOptions function

### ISSUE THREAD

**Issue by chanon** [vuejs/vue#4976]
While investigating the cause of https://github.com/vuejs/vue-class-component/issues/65 related to `vue-hot-reload` not working with `vue-class-component`, @ktsn found a bug in vue core which he noted here: 
https://github.com/vuejs/vue-class-component/issues/65#issuecomment-280859079

He says

> I find there is a bug in core library that accidentally drop late-injected options of constructors. That means, if we use vue-hot-reload-api or vue-loader, they inject some options into component options object after creating component constructors, then the component is instantiated by using constructor with $createElement and the injected options are dropped in [resolveComponentOptions function](https://github.com/vuejs/vue/blob/dev/src/core/instance/init.js#L69). 

And that it is also causing https://github.com/vuejs/vue-class-component/issues/66

Here is the minimal reproduction of this bug by @ktsn 
https://jsfiddle.net/vvxLyLvq/2/

I don't know much about it, I'm just creating the issue here for @ktsn 

In summary this issue is causing `vue-class-component` hot-reload to not work and also makes using `vue-class-component` with `CSS Modules` impossible.

Thank you!
